src/v/raft/replication_monitor.cc

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "raft/replication_monitor.h"
+
+#include "raft/consensus.h"
+#include "raft/errc.h"
+
+namespace raft {
+
+replication_monitor::replication_monitor(consensus* r)
+  : _raft(r) {
+    ssx::repeat_until_gate_closed(_gate, [this] {
+        return do_notify_replicated().handle_exception(
+          [this](const std::exception_ptr& e) {
+              if (!ssx::is_shutdown_exception(e)) {
+                  vlog(
+                    _raft->_ctxlog.error,
+                    "Exception running replication monitor, ignoring: {}",
+                    e);
+              }
+          });
+    });
+
+    ssx::repeat_until_gate_closed(_gate, [this] {
+        return do_notify_committed().handle_exception(
+          [this](const std::exception_ptr& e) {
+              if (!ssx::is_shutdown_exception(e)) {
+                  vlog(
+                    _raft->_ctxlog.error,
+                    "Exception running replication monitor, ignoring: {}",
+                    e);
+              }
+          });
+    });
+}
+
+std::ostream&
+operator<<(std::ostream& o, const replication_monitor::waiter& entry) {
+    fmt::print(o, "{{type: {}, append result: {}}}", entry.type, entry.result);
+    return o;
+}
+
+std::ostream&
+operator<<(std::ostream& o, const replication_monitor::wait_type& type) {
+    switch (type) {
+    case replication_monitor::commit:
+        fmt::print(o, "commit");
+        break;
+    case replication_monitor::majority_replication:
+        fmt::print(o, "majority_replication");
+        break;
+    }
+    return o;
+}
+
+ss::future<> replication_monitor::stop() {
+    auto f = _gate.close();
+    for (auto& [_, waiter] : _waiters) {
+        waiter.done.set_value(errc::shutting_down);
+    }
+    _committed_event_cv.broken();
+    _replicated_event_cv.broken();
+    return f;
+}
+
+ss::future<errc> replication_monitor::do_wait_until(
+  storage::append_result append, wait_type type) {
+    if (_gate.is_closed()) {
+        return ssx::now(errc::not_leader);
+    }
+    auto it = _waiters.emplace(append.base_offset, waiter(append, type));
+    return it->second.done.get_future();
+}
+
+ss::future<errc>
+replication_monitor::wait_until_committed(storage::append_result append) {
+    return do_wait_until(append, wait_type::commit);
+}
+
+ss::future<errc> replication_monitor::wait_until_majority_replicated(
+  storage::append_result append) {
+    return do_wait_until(append, wait_type::majority_replication);
+}
+
+ss::future<> replication_monitor::do_notify_replicated() {
+    while (!_gate.is_closed()) {
```

Comment:
do we need two loops here ? it is already being repeated in the ` ssx::repeat_until_gate_closed`